include/seahorn/CexExeGenerator.hh

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/ValueMap.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+#include "boost/algorithm/string/replace.hpp"
+
+#include "seahorn/Bmc.hh"
+#include "seahorn/Expr/ExprLlvm.hh"
+#include "seahorn/Expr/ExprMemMap.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+#include "seahorn/MemSimulator.hh"
+#include "seahorn/SolverBmc.hh"
+#include "seahorn/Support/SeaDebug.h"
+#include "seahorn/Transforms/Instrumentation/ShadowMemDsa.hh"
+#include <memory>
+
+namespace llvm {
+class TargetLibraryInfo;
+class DataLayout;
+class LLVMContext;
+class Module;
+} // namespace llvm
+
+/*
+    Converts bmc trace into a linkable LLVM IR
+    with all non-det functions implemented
+ */
+
+namespace seahorn {
+namespace cexGen {
+
+using MemMap = expr::exprMemMap::ExprMemMap;
+using DsaKeyVal = std::map<Expr, Expr>;
+
+// helper functions
+namespace utils {
+
+/// \brief converts Expr \p e into LLVM Constant
+Constant *exprToLlvm(Type *ty, Expr e, LLVMContext &ctx, const DataLayout &dl);
+
+/// \brief extract default value and special id-val pairs from Expr encoding of
+/// a shadow mem segment \p segment using ExprMemMap; returns a ConstantArray
+/// of length \p size divided by content width, filled with default and special
+/// values
+Constant *exprToMemSegment(Expr segment, Expr startAddr, Expr size,
```

Comment:
again, move into some class to simplify the args at each call. same class as prev method